MAVProxy/modules/mavproxy_paraedit/param_editor.py

```diff
@@ -0,0 +1,212 @@
+
+    def run(self):
+        while not self.time_to_quit:
+            queue_access_start_time = time.time()
+            self.event_queue_lock.acquire()
+            request_read_after_processing_queue = False
+            while (self.event_queue.qsize() > 0 and
+                    (time.time() - queue_access_start_time) < 0.6):
+                event = self.event_queue.get()
+                event_type = event.get_type()
+
+                if event_type == ph_event.PEE_READ_PARAM:
+                    self.param_received = self.module('param').mav_param
+                    self.mp_paraedit.gui_event_queue_lock.acquire()
+                    self.mp_paraedit.gui_event_queue.put(ParamEditorEvent(
+                        ph_event.PEGE_READ_PARAM, param=self.param_received))
+                    self.mp_paraedit.gui_event_queue_lock.release()
+
+                elif event_type == ph_event.PEE_LOAD_FILE:
+                    self.module('param').mav_param.load(
+                            event.get_arg("path"), "*")
+
+                elif event_type == ph_event.PEE_SAVE_FILE:
+                    self.module('param').mav_param.save(
+                            event.get_arg("path"), "*", verbose=True)
+
+                elif event_type == ph_event.PEE_WRITE_PARAM:
+                    paramchanged = event.get_arg("modparam")
+                    for param, value in paramchanged.items():
+                        time.sleep(1)
+                        sysid = self.module('param').get_sysid()
+                        setparam = []
+                        setparam.append('set')
+                        setparam.append(param)
+                        setparam.append(value)
+                        self.module('param').pstate[sysid].handle_command(
+                            self.master, self.mp_paraedit.mpstate, setparam)
+
+                    # don't modify queue while in the middile of processing it:
+                    request_read_after_processing_queue = True
+
+            self.event_queue_lock.release()
+
+            # if event processing operations require a mission referesh in GUI
+            # (e.g., after a load or a verified-completed write):
+            if (request_read_after_processing_queue):
+                self.event_queue_lock.acquire()
+                self.event_queue.put(ParamEditorEvent(ph_event.PEE_READ_PARAM))
+                self.event_queue_lock.release()
+
+            # periodically re-request WPs that were never received:
+            # DON'T NEED TO! -- wp module already doing this
+
+            time.sleep(0.2)
+
+
+class ParamEditorMain(object):
+    def __init__(self, mpstate):
+        self.param_received = {}
+
+        self.event_queue = multiproc.Queue()
+        self.event_queue_lock = multiproc.Lock()
+        self.gui_event_queue = multiproc.Queue()
+        self.gui_event_queue_lock = multiproc.Lock()
+
+        self.close_window = multiproc.Semaphore()
```

Comment:
Have you tested closing the window unloads the module and you can then reload the module and get the gui back?

Reply:
Made some changes. It works fine now. Though its a common issue with many modules that the module doesnt get unloaded on closing the window. Would be better to solve it.